gnoi/os/server.go

```diff
@@ -51,3 +61,118 @@ func (s *Server) Activate(ctx context.Context, request *pb.ActivateRequest) (*pb
 	}
 	return &pb.ActivateResponse{Response: &pb.ActivateResponse_ActivateOk{}}, nil
 }
+
+// Install receives an OS package, validates the package and then installs the package.
+func (s *Server) Install(stream pb.OS_InstallServer) error {
+	var request *pb.InstallRequest
+	var response *pb.InstallResponse
+	var err error
+	if request, err = stream.Recv(); err != nil {
+		return err
+	}
+	utils.LogProto(request)
+	transferRequest := request.GetTransferRequest()
+	if transferRequest == nil {
+		return errors.New("Failed to receive TransferRequest")
+	}
+	if s.manager.InstallInProgress() {
+		response = &pb.InstallResponse{Response: &pb.InstallResponse_InstallError{InstallError: &pb.InstallError{Type: pb.InstallError_INSTALL_IN_PROGRESS}}}
+		utils.LogProto(response)
+		stream.Send(response)
+		return errors.New("Another install is already in progress")
+	}
+	if s.manager.IsRunning(transferRequest.Version) {
+		response = &pb.InstallResponse{Response: &pb.InstallResponse_InstallError{
+			InstallError: &pb.InstallError{Type: pb.InstallError_INSTALL_RUN_PACKAGE},
+		}}
+		utils.LogProto(response)
+		stream.Send(response)
+		return errors.New("Attempting to force transfer an OS of the same version as the currently running OS")
+	}
+	if version := transferRequest.Version; s.manager.IsInstalled(version) {
+		response = &pb.InstallResponse{Response: &pb.InstallResponse_Validated{
+			Validated: &pb.Validated{
+				Version: version,
+			},
+		}}
+		utils.LogProto(response)
+		if err = stream.Send(response); err != nil {
+			return err
+		}
+		return nil
+	}
+	s.manager.SetInstallInProgress(true)
+	defer s.manager.SetInstallInProgress(false)
+
+	response = &pb.InstallResponse{Response: &pb.InstallResponse_TransferReady{TransferReady: &pb.TransferReady{}}}
+	utils.LogProto(response)
+	if err = stream.Send(response); err != nil {
+		return err
+	}
+	errorChan := make(chan error)
+	updateProgress := make(chan uint64)
+	transferredOS := make(chan *bytes.Buffer)
+	go ReceiveOS(stream, errorChan, updateProgress, transferredOS)
```

Comment:
=) You look like me the first time I played with channels. Let me ask you a simple question and perhaps that will lead you to simplify this code. Why do you need a go thread here? ;) At this point all you need to do is, receive data, receive transfer end or handle a stream error.

Reply:
- receive TransferRequest
- check version is the one running
- check if another install is already in progress, else quit with error
- lock this as being the only install, all others will fail
- received next response (transfer content or transfer end or error)
- if transfer content totals an amount of chunks then log the amount
- if error, then finish
- if transfer end then validate image and return success
- release the lock

Reply:
forgot to include the "send the sync progress" wich you can interleave in between every reception.